### PR TITLE
Add queue for cold entities

### PR DESCRIPTION
**Pull request by kaituo**
Note: since there are a lot of dependencies, I only list the main class and test code to save reviewers' time. The build will fail due to missing dependencies. I will use that PR just for review. will not merge it. Will have a big one in the end and merge once after all review PRs get approved.  Now the code is missing unit tests.  Posting PRs now to meet the cutoff date (June 1).  Will add unit tests, run performance tests, and fix bugs before the official release.

### Description
In HCAD v1, only a subset of hot entities can use their models to predict if we are short of memory. In v2, we allocate a queue to absorb cold entities. Like hot entities, we load a cold entity's model checkpoint from disk, train models if the checkpoint is not found, query for missed features to complete a shingle, use the models to check whether the incoming feature is normal, update models, and save the detection results to disks.

This PR adds the implementation of cold entity queue. Implementation-wise, we reuse the queues we have developed for hot entities.  The differences are we release cold entity requests in a controlled pace.  Also, cold entity requests’ priority is low.  So only when there are no hot entities requests to process we are gonna process cold entity requests.

Testing done:
1. Manual tests using 10 HCAD detectors and 12,000 entities in a 3 node cluster.

### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ Y ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+ * thread pool has availability) are available, while we release cold entity requests
+ * to other queues at a slow controlled pace. Also, cold entity requests' priority is low.
+ * So only when there are no hot entity requests to process are we going to process cold
+ * entity requests. 
+ *
+ */
+public class ColdEntityQueue extends RateLimitedQueue<EntityFeatureRequest> {
+    private static final Logger LOG = LogManager.getLogger(ColdEntityQueue.class);
+
+    private int batchSize;
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:67]: Add `volatile`?

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    private final CheckpointReadQueue checkpointReadQueue;
+    private boolean scheduled;
+    private int expectedExecutionTimeInSecsPerRequest;
+
+    public ColdEntityQueue(
+        long heapSizeInBytes,
+        int singleRequestSizeInBytes,
+        Setting<Float> maxHeapPercentForQueueSetting,
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:75]: Why not get this setting like line 109 ?

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+ * So only when there are no hot entity requests to process are we going to process cold
+ * entity requests. 
+ *
+ */
+public class ColdEntityQueue extends RateLimitedQueue<EntityFeatureRequest> {
+    private static final Logger LOG = LogManager.getLogger(ColdEntityQueue.class);
+
+    private int batchSize;
+    private final CheckpointReadQueue checkpointReadQueue;
+    private boolean scheduled;
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:69]: What does `scheduled` mean? Add some comments?

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            }
+        }
+        return requestSize;
+    }
+
+    private synchronized void schedulePulling(TimeValue delay) {
+        try {
+            threadPool.schedule(this::pullRequests, delay, AnomalyDetectorPlugin.AD_THREAD_POOL_NAME);
+        } catch (Exception e) {
+            LOG.error("Fail to schedule cold entity pulling", e);
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:151]: What will happen if fail to schedule? Seems we will just poll part of requests and put into `checkpointReadQueue` and ignore other requests if fail to schedule.

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            pullRequests();
+        }
+    }
+
+    /**
+     * @param requestSize requests to process
+     * @return the delay for the next scheduled run
+     */
+    private TimeValue getScheduleDelay(int requestSize) {
+        int expectedSingleRequestExecutionMillis = 1000 * expectedExecutionTimeInSecsPerRequest;
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:170]: Could you add a comment here explaining what these calculations are achieving? For instance, why do we wait longer if the requestSize is larger?

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+ * A queue slowly releasing low-priority requests to CheckpointReadQueue
+ *
+ * ColdEntityQueue is a queue to absorb cold entities. Like hot entities, we load a cold
+ * entity's model checkpoint from disk, train models if the checkpoint is not found,
+ * query for missed features to complete a shingle, use the models to check whether
+ * the incoming feature is normal, update models, and save the detection results to disks. 
+ * Implementation-wise, we reuse the queues we have developed for hot entities.
+ * The differences are: we process hot entities as long as resources (e.g., AD
+ * thread pool has availability) are available, while we release cold entity requests
+ * to other queues at a slow controlled pace. Also, cold entity requests' priority is low.
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:59]: To clarify, can this queue only have low priority requests?

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:67]: added

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+/*
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:12]: right, removed

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java (hunk 6 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:59]: yes

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java (hunk 3 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:69]: added comment "// indicate whether a future pull over cold entity queues is scheduled"

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:75]: I did it in the super class to reduce redundant code.

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        Clock clock,
+        float mediumSegmentPruneRatio,
+        float lowSegmentPruneRatio,
+        int maintenanceFreqConstant,
+        CheckpointReadQueue checkpointReadQueue,
+        Duration stateTtl,
+        NodeStateManager nodeStateManager
+    ) {
+        super(
+            "cold-entity",
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:91]: It is only used here.  Will do it when needed.

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java (hunk 4 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:151]: I am logging failures.  threadpool.schedule is the basic API provided by Opensearch.  If it fails, I don't have fallback as sth is fundamentally wrong.

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java (hunk 5 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java:170]: added the following comments.

/**
     * The method calculates the delay we have to set to control the rate of cold
     * entity processing. We wait longer if the requestSize is larger to give the
     * system more time to processing requests.  We ddd randomness to cope with the
     * case that we want to execute at least 1 request every few seconds, but
     * cannot guarantee that.
     * @param requestSize requests to process
     * @return the delay for the next scheduled run
     */

File: src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java (hunk 6 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java]: How is it guaranteed that it only has low priority requests? Isnt it possible for someone to add a high priority request to this queue? There aren't any safe guards, correct?

Comment (kaituo) [reply src/main/java/org/opensearch/ad/ratelimit/ColdEntityQueue.java]: I added an extra filtering to safeguard that.  Please check new commits.

**Review by jmazanec15 — APPROVED** [review]

**Review by jngz-es — APPROVED** [review]

**Pull request closed without merging**